From 65f53b85d4ac88df15333172e4d8b10fec2c9906 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Tue, 14 Nov 2023 16:49:26 +0100
Subject: [PATCH 0864/1013] evl/net: socket: add low-level memory charge
 routine

We may need to charge memory which is not carried by a skb. Provide an
inner interface for this purpose, basing the original skb-based call
on it.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/skb.h          | 12 ++++++++++
 include/evl/net/socket.h       | 14 ++++++------
 kernel/evl/net/output.c        |  4 ++--
 kernel/evl/net/packet/packet.c | 10 ++++-----
 kernel/evl/net/skb.c           | 40 ++++++++++++++++++++++++++++++++++
 kernel/evl/net/socket.c        | 28 +++++++-----------------
 6 files changed, 74 insertions(+), 34 deletions(-)

diff --git a/include/evl/net/skb.h b/include/evl/net/skb.h
index 8c92fff0377d..44ae1d6a1890 100644
--- a/include/evl/net/skb.h
+++ b/include/evl/net/skb.h
@@ -56,6 +56,18 @@ void evl_net_free_skb_list(struct list_head *list);
 
 struct sk_buff *evl_net_clone_skb(struct sk_buff *skb);
 
+bool evl_net_charge_skb_rmem(struct evl_socket *esk,
+			struct sk_buff *skb);
+
+void evl_net_uncharge_skb_rmem(struct evl_socket *esk,
+			struct sk_buff *skb);
+
+int evl_net_charge_skb_wmem(struct evl_socket *esk,
+			struct sk_buff *skb,
+			ktime_t timeout, enum evl_tmode tmode);
+
+void evl_net_uncharge_skb_wmem(struct sk_buff *skb);
+
 int evl_net_init_pools(void);
 
 ssize_t evl_net_show_clones(char *buf, size_t len);
diff --git a/include/evl/net/socket.h b/include/evl/net/socket.h
index 98247023f668..334101b82319 100644
--- a/include/evl/net/socket.h
+++ b/include/evl/net/socket.h
@@ -88,29 +88,29 @@ static inline unsigned int evl_socket_f_flags(struct evl_socket *esk)
 }
 
 static inline bool evl_charge_socket_rmem(struct evl_socket *esk,
-					  struct sk_buff *skb)
+					  size_t size)
 {
 	/* An overflow of skb->truesize - 1 is allowed, not more. */
 	if (atomic_read(&esk->rmem_count) >= esk->rmem_max)
 		return false;
 
 	/* We don't have to saturate, atomic_t is fine. */
-	atomic_add(skb->truesize, &esk->rmem_count);
+	atomic_add(size, &esk->rmem_count);
 
 	return true;
 }
 
 static inline void evl_uncharge_socket_rmem(struct evl_socket *esk,
-					    struct sk_buff *skb)
+					    size_t size)
 {
-	atomic_sub(skb->truesize, &esk->rmem_count);
+	int count = atomic_sub_return(size, &esk->rmem_count);
+	EVL_WARN_ON(NET, count < 0);
 }
 
-int evl_charge_socket_wmem(struct evl_socket *esk,
-			struct sk_buff *skb,
+int evl_charge_socket_wmem(struct evl_socket *esk, size_t size,
 			ktime_t timeout, enum evl_tmode tmode);
 
-void evl_uncharge_socket_wmem(struct sk_buff *skb);
+void evl_uncharge_socket_wmem(struct evl_socket *esk, size_t size);
 
 int evl_register_socket_domain(struct evl_socket_domain *domain);
 
diff --git a/kernel/evl/net/output.c b/kernel/evl/net/output.c
index 22ccd2a56ec0..70b514536288 100644
--- a/kernel/evl/net/output.c
+++ b/kernel/evl/net/output.c
@@ -40,7 +40,7 @@ oob_start_xmit(struct net_device *dev, struct sk_buff *skb)
 static inline void do_tx(struct evl_net_qdisc *qdisc,
 			struct net_device *dev, struct sk_buff *skb)
 {
-	evl_uncharge_socket_wmem(skb);
+	evl_net_uncharge_skb_wmem(skb);
 
 	switch (oob_start_xmit(dev, skb)) {
 	case NETDEV_TX_OK:
@@ -103,7 +103,7 @@ void evl_net_do_tx(void *arg)
 
 static void skb_xmit_inband(struct sk_buff *skb)
 {
-	evl_uncharge_socket_wmem(skb);
+	evl_net_uncharge_skb_wmem(skb);
 	skb->prev = NULL;
 	skb->next = NULL;
 	dev_queue_xmit(skb);
diff --git a/kernel/evl/net/packet/packet.c b/kernel/evl/net/packet/packet.c
index 5824a0a0d5e5..ed78129c45df 100644
--- a/kernel/evl/net/packet/packet.c
+++ b/kernel/evl/net/packet/packet.c
@@ -90,7 +90,7 @@ static bool __packet_deliver(struct evl_net_rxqueue *rxq,
 		 * not consume more memory, skip delivery and try with
 		 * the next subscriber.
 		 */
-		if (!evl_charge_socket_rmem(esk, skb))
+		if (!evl_net_charge_skb_rmem(esk, skb))
 			continue;
 
 		/*
@@ -104,7 +104,7 @@ static bool __packet_deliver(struct evl_net_rxqueue *rxq,
 			qskb = evl_net_clone_skb(skb);
 			if (qskb == NULL) {
 				evl_flush_wait(&esk->input_wait, EVL_T_NOMEM);
-				evl_uncharge_socket_rmem(esk, skb);
+				evl_net_uncharge_skb_rmem(esk, skb);
 				break;
 			}
 		}
@@ -489,13 +489,13 @@ static ssize_t send_packet(struct evl_socket *esk,
 	 * output contention to end, or the caller asked for a
 	 * non-blocking operation while such contention was ongoing.
 	 */
-	ret = evl_charge_socket_wmem(esk, skb, timeout, tmode);
+	ret = evl_net_charge_skb_wmem(esk, skb, timeout, tmode);
 	if (ret)
 		goto cleanup;
 
 	ret = evl_net_ether_transmit(dev, skb);
 	if (ret) {
-		evl_uncharge_socket_wmem(skb);
+		evl_net_uncharge_skb_wmem(skb);
 		goto cleanup;
 	}
 
@@ -611,7 +611,7 @@ static ssize_t receive_packet(struct evl_socket *esk,
 			/* Restore the MAC header. */
 			skb_push(skb, skb->data - skb_mac_header(skb));
 			ret = copy_packet_to_user(u_msghdr, iov, iovlen, skb);
-			evl_uncharge_socket_rmem(esk, skb);
+			evl_net_uncharge_skb_rmem(esk, skb);
 			evl_net_free_skb(skb);
 			return ret;
 		}
diff --git a/kernel/evl/net/skb.c b/kernel/evl/net/skb.c
index 391c0dd1940c..19ee8c2ab021 100644
--- a/kernel/evl/net/skb.c
+++ b/kernel/evl/net/skb.c
@@ -16,6 +16,7 @@
 #include <evl/list.h>
 #include <evl/work.h>
 #include <evl/wait.h>
+#include <evl/net/socket.h>
 
 static unsigned int net_clones = 1024; /* FIXME: Kconfig? */
 
@@ -406,6 +407,45 @@ static struct sk_buff *alloc_one_skb(struct net_device *dev)
 	return skb;
 }
 
+bool evl_net_charge_skb_rmem(struct evl_socket *esk, struct sk_buff *skb)
+{
+	return evl_charge_socket_rmem(esk, skb->truesize);
+}
+
+void evl_net_uncharge_skb_rmem(struct evl_socket *esk, struct sk_buff *skb)
+{
+	evl_uncharge_socket_rmem(esk, skb->truesize);
+}
+
+int evl_net_charge_skb_wmem(struct evl_socket *esk,
+			struct sk_buff *skb,
+			ktime_t timeout, enum evl_tmode tmode)
+{
+	int ret;
+
+	EVL_NET_CB(skb)->tracker = NULL;
+	ret = evl_charge_socket_wmem(esk, skb->truesize, timeout, tmode);
+	if (!ret)
+		EVL_NET_CB(skb)->tracker = esk;
+
+	return ret;
+}
+
+void evl_net_uncharge_skb_wmem(struct sk_buff *skb)
+{
+	struct evl_socket *esk = EVL_NET_CB(skb)->tracker;
+
+	/*
+	 * If set, the tracking socket cannot be stale as it has to
+	 * pass the wmem_crossing first before unwinding in
+	 * sock_oob_detach().
+	 */
+	if (esk) {
+		EVL_NET_CB(skb)->tracker = NULL;
+		evl_uncharge_socket_wmem(esk, skb->truesize);
+	}
+}
+
 /* in-band */
 int evl_net_dev_build_pool(struct net_device *dev)
 {
diff --git a/kernel/evl/net/socket.c b/kernel/evl/net/socket.c
index e313c79c8514..11bc0de7f6c8 100644
--- a/kernel/evl/net/socket.c
+++ b/kernel/evl/net/socket.c
@@ -145,51 +145,39 @@ void evl_unregister_socket_domain(struct evl_socket_domain *domain)
 	mutex_unlock(&domain_lock);
 }
 
-static inline bool charge_socket_wmem(struct evl_socket *esk,
-				struct sk_buff *skb)
+static inline bool charge_socket_wmem(struct evl_socket *esk, size_t size)
 {				/* esk->wmem_wait.wchan.lock held */
 	if (atomic_read(&esk->wmem_count) >= esk->wmem_max)
 		return false;
 
-	atomic_add(skb->truesize, &esk->wmem_count);
-	EVL_NET_CB(skb)->tracker = esk;
+	atomic_add(size, &esk->wmem_count);
 	evl_down_crossing(&esk->wmem_drain);
 
 	return true;
 }
 
-int evl_charge_socket_wmem(struct evl_socket *esk,
-			struct sk_buff *skb,
-			ktime_t timeout, enum evl_tmode tmode)
+int evl_charge_socket_wmem(struct evl_socket *esk, size_t size,
+		ktime_t timeout, enum evl_tmode tmode)
 {
-	EVL_NET_CB(skb)->tracker = NULL;
-
 	if (!esk->wmem_max)	/* Unlimited. */
 		return 0;
 
-	return evl_wait_event_timeout(&esk->wmem_wait,
-				timeout, tmode,
-				charge_socket_wmem(esk, skb));
+	return evl_wait_event_timeout(&esk->wmem_wait, timeout, tmode,
+				charge_socket_wmem(esk, size));
 }
 
-void evl_uncharge_socket_wmem(struct sk_buff *skb)
+void evl_uncharge_socket_wmem(struct evl_socket *esk, size_t size)
 {
-	struct evl_socket *esk = EVL_NET_CB(skb)->tracker;
 	unsigned long flags;
 	int count;
 
-	if (!esk)
-		return;
-
 	/*
 	 * The tracking socket cannot be stale as it has to pass the
 	 * wmem_crossing first before unwinding in sock_oob_detach().
 	 */
 	raw_spin_lock_irqsave(&esk->wmem_wait.wchan.lock, flags);
 
-	EVL_NET_CB(skb)->tracker = NULL;
-
-	count = atomic_sub_return(skb->truesize, &esk->wmem_count);
+	count = atomic_sub_return(size, &esk->wmem_count);
 	if (count < esk->wmem_max && evl_wait_active(&esk->wmem_wait))
 		evl_flush_wait_locked(&esk->wmem_wait, 0);
 
-- 
2.30.2

